### PR TITLE
SWATCH-839: Add new Admin Api for Contract Sync

### PR DESCRIPTION
**Pull request by ksynvrit**
This is to create the Admin API to sync existing contracts and add new contracts if there are not already in the swatch-contract database.

https://issues.redhat.com/browse/SWATCH-839

Testing Steps:

1. Run with with these environment variables:
```
SWATCH_SELF_PSK=placeholder;KEYSTORE_RESOURCE=file:/home/<user>/IdeaProjects/rhsm-subscriptions/config/certs/keystore.jks;ENABLE_SPLUNK_HEC=false;KEYSTORE_PASSWORD=<password>;QUARKUS_PROFILE=stage
```
If you have problems access the Partner API you may need to pull credentials from vault [here](https://docs.engineering.redhat.com/pages/viewpage.action?spaceKey=ENT&title=Setting+up+non-prod+credentials)
2. Setup database with test data (both contracts and contract-metrics):
```
INSERT INTO public.contracts (uuid, subscription_number, last_updated, start_date, end_date, org_id, sku, billing_provider, billing_account_id, product_id) VALUES ('6c849fcb-c59f-41b0-9e82-11d3a33898e9', '12400374', '2023-03-16 21:30:43.808568 +00:00', '2023-03-16 21:30:10.414684 +00:00', '2023-03-16 21:30:43.808568 +00:00', '13468036', 'MW01484', 'aws_marketplace', '896801664647', 'OpenShift-dedicated-metrics');
INSERT INTO public.contracts (uuid, subscription_number, last_updated, start_date, end_date, org_id, sku, billing_provider, billing_account_id, product_id) VALUES ('eaa10ad4-4466-47d3-a3f4-f003e924a02b', '12400374', '2023-03-16 21:30:52.473337 +00:00', '2023-03-16 21:30:52.473337 +00:00', null, '13468036', 'MW01484', 'aws_marketplace', '896801664647', 'OpenShift-dedicated-metrics');
INSERT INTO public.contracts (uuid, subscription_number, last_updated, start_date, end_date, org_id, sku, billing_provider, billing_account_id, product_id) VALUES ('9bf28398-d147-4a4d-97f1-2e3ab46af681', 'subscription123', '2023-03-23 19:57:41.369450 +00:00', '2023-01-01 00:00:00.000000 +00:00', '2023-02-01 00:00:00.000000 +00:00', 'org123', 'BASILISK', 'aws', '123456', 'BASILISK');
INSERT INTO public.contracts (uuid, subscription_number, last_updated, start_date, end_date, org_id, sku, billing_provider, billing_account_id, product_id) VALUES ('7e5f1aa2-0805-42df-8254-097efd98cfbd', 'subscription123', '2023-03-23 19:57:41.369450 +00:00', '2023-02-01 00:00:00.000000 +00:00', null, 'org123', 'BASILISK', 'aws', '123456', 'BASILISK');
INSERT INTO public.contracts (uuid, subscription_number, last_updated, start_date, end_date, org_id, sku, billing_provider, billing_account_id, product_id) VALUES ('642a3b85-240a-4c1f-ab4e-3ee660935a64', '12400374', '2023-03-16 21:30:52.473337 +00:00', '2023-03-16 21:30:43.808568 +00:00', '2023-03-16 21:30:52.473337 +00:00', '13468036', 'MW01484', 'aws_marketplace', '896801664647', 'OpenShift-dedicated-metrics');
```
```
INSERT INTO public.contract_metrics (contract_uuid, metric_id, metric_value) VALUES ('6c849fcb-c59f-41b0-9e82-11d3a33898e9', 'test_dim_1', 5);
INSERT INTO public.contract_metrics (contract_uuid, metric_id, metric_value) VALUES ('6c849fcb-c59f-41b0-9e82-11d3a33898e9', 'test_dim_2', 10);
INSERT INTO public.contract_metrics (contract_uuid, metric_id, metric_value) VALUES ('642a3b85-240a-4c1f-ab4e-3ee660935a64', 'test_dim_3', 5);
INSERT INTO public.contract_metrics (contract_uuid, metric_id, metric_value) VALUES ('642a3b85-240a-4c1f-ab4e-3ee660935a64', 'test_dim_2', 10);
INSERT INTO public.contract_metrics (contract_uuid, metric_id, metric_value) VALUES ('eaa10ad4-4466-47d3-a3f4-f003e924a02b', 'test_dim_3', 10);
INSERT INTO public.contract_metrics (contract_uuid, metric_id, metric_value) VALUES ('eaa10ad4-4466-47d3-a3f4-f003e924a02b', 'test_dim_2', 10);
```
3. Run `syncContractByOrgId` & `syncAllContracts` in swatch-contract UI:
`http://localhost:8000/q/swagger-ui/#/default/syncContractsByOrg`for UI enter `13468036` or curl command:
```
curl -X 'POST' \
  'http://localhost:8000/api/swatch-contracts/internal/rpc/sync/contracts/13468036' \
  -H 'accept: application/json' \
  -H 'x-rh-swatch-psk: placeholder' \
  -d ''
```
Observe that the status gives you the message that the sync was succuessful & that the Databases is updated in `contract` table.

### CODE REVIEW

File: swatch-contracts/src/main/java/com/redhat/swatch/contract/resource/ContractsTestingResource.java [hunk 1]
```diff
@@ -118,6 +119,24 @@ public Contract updateContract(String uuid, Contract contract)
     return service.updateContract(contract);
   }
 
+  @Override
+  @Transactional
+  @RolesAllowed({"test"})
+  public StatusResponse syncAllContracts() throws ApiException, ProcessingException {
+    log.info("Syncing All Contracts");
+    var contracts = service.getAllContracts();
+    for (ContractEntity org : contracts) {
```
Comment (kartikshahc) [inline swatch-contracts/src/main/java/com/redhat/swatch/contract/resource/ContractsTestingResource.java:128]: If the contracts list is empty then show the message "No active contract found for the orgIds"

**Review by kartikshahc — CHANGES_REQUESTED** [review]

File: swatch-contracts/src/main/java/com/redhat/swatch/contract/service/ContractService.java [hunk 2]
```diff
@@ -284,6 +287,57 @@ public StatusResponse createPartnerContract(PartnerEntitlementContract contract)
...
+                    .source(contract.getBillingProvider()));
+        if (Objects.nonNull(result.getContent()) && !result.getContent().isEmpty()) {
+          var entitlementEntity =
+              transformEntitlementToContractEntity(result.getContent(), contract);
+          var isDuplicateContract = isDuplicateContract(entitlementEntity, contract);
+          if (Objects.nonNull(entitlementEntity)) {
+            if (isDuplicateContract) {
+              log.info("Syncing Existing Contract for {}", contractOrgSync);
+              statusResponse.setStatus("SUCCESS");
+              persistExistingContract(entitlementEntity, OffsetDateTime.now());
```
[the remark this answers is not in the document]
Comment (ksynvrit) [reply swatch-contracts/src/main/java/com/redhat/swatch/contract/service/ContractService.java:313]: I was using this to update existing records

File: swatch-contracts/src/main/java/com/redhat/swatch/contract/service/ContractService.java [hunk 3]
```diff
@@ -326,10 +380,55 @@ private Optional<ContractEntity> currentlyActiveContract(ContractEntity contract
...
+
+        var rhEntitlements = entitlement.getRhEntitlements();
+        if (Objects.nonNull(rhEntitlements)
+            && !rhEntitlements.isEmpty()
+            && Objects.nonNull(rhEntitlements.get(0))) {
+          var subscription = rhEntitlements.get(0).getRedHatSubscriptionNumber();
+          var sku = rhEntitlements.get(0).getSku();
+          currentEntity.setSku(sku);
+          currentEntity.setSubscriptionNumber(subscription);
+          OfferingProductTags productTags = syncResource.getSkuProductTags(sku);
```
[the remark this answers is not in the document]
Comment (ksynvrit) [reply swatch-contracts/src/main/java/com/redhat/swatch/contract/service/ContractService.java:418]: I decide to use the `throw` declaration instead here. Let me know if your find with that

File: swatch-contracts/src/main/java/com/redhat/swatch/contract/service/ContractService.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (kartikshahc) [reply swatch-contracts/src/main/java/com/redhat/swatch/contract/service/ContractService.java:321]: If its same record then I don't think we need to update. Unless you can think of a reason?

**Review by kartikshahc — CHANGES_REQUESTED** [review]

**Review by kartikshahc — APPROVED** [review]
LGTM

**Pull request merged**